File: packages/ide-adapters/vscode/src/services/coreBridge.ts
import * as vscode from 'vscode';
import * as path from 'path';
import * as fs from 'fs';
import { spawn, ChildProcess } from 'child_process';
import * as readline from 'readline';

const DEFAULT_TIMEOUT_MS = 10000;
const MAX_CRASHES = 3;
const DISPOSE_TIMEOUT_MS = 2000;

/** Response object returned by the .NET Core process. */
export interface CoreResponse {
  success: boolean;
  result: string;
  error: string;
}

/** Result of a syntax validation check. */
export interface ValidationResult {
  isValid: boolean;
  diagnostics: DiagnosticEntry[];
}

/** A single diagnostic entry from syntax validation. */
export interface DiagnosticEntry {
  severity: string;
  message: string;
  line: number;
  column: number;
}

/** Bridge to the .NET Core translation engine using a persistent long-lived process. */
export class CoreBridge {
  public coreDllPath: string;
  public translationsPath: string;
  public projectPath: string;
  public outputChannel: vscode.OutputChannel;
  public timeoutMs: number;

  private process: ChildProcess | null = null;
  private responseReader: readline.Interface | null = null;
  private pendingResolve: ((line: string) => void) | null = null;
  private requestQueue: Promise<void> = Promise.resolve();
  private crashCount: number = 0;
  private consecutiveTimeouts: number = 0;
  private disposed: boolean = false;

  constructor(
    context: vscode.ExtensionContext,
    outputChannel: vscode.OutputChannel,
    timeoutMs: number = DEFAULT_TIMEOUT_MS
  ) {
    this.coreDllPath = path.join(
      context.extensionPath,
      'bin',
      'MultiLingualCode.Core.Host.dll'
    );
    this.projectPath = '';
    this.outputChannel = outputChannel;
    this.timeoutMs = timeoutMs;

    const workspaceFolders: readonly vscode.WorkspaceFolder[] | undefined =
      vscode.workspace.workspaceFolders;
    if (workspaceFolders && workspaceFolders.length > 0) {
      this.projectPath = workspaceFolders[0].uri.fsPath;
    }

    this.translationsPath = this.resolveTranslationsPath(context);
  }

  /** Starts the persistent .NET Core process if not already running. */
  public startProcess(): void {
    if (this.process) {
      return;
    }

    const args: string[] = [this.coreDllPath];

    if (this.translationsPath) {
      args.push('--translations', this.translationsPath);
    }
    if (this.projectPath) {
      args.push('--project', this.projectPath);
    }

    this.outputChannel.appendLine('CoreBridge: starting persistent process');
    const coreProcess: ChildProcess = spawn('dotnet', args);

    coreProcess.on('exit', (code: number | null): void => {
      this.outputChannel.appendLine(`CoreBridge: process exited with code ${code}`);
      this.process = null;
      this.responseReader = null;
      this.pendingResolve = null;

      if (!this.disposed) {
        this.crashCount++;
      }
    });

    coreProcess.on('error', (err: Error): void => {
      this.outputChannel.appendLine(`CoreBridge: process error - ${err.message}`);
      this.process = null;
      this.responseReader = null;
      this.pendingResolve = null;

      if (!this.disposed) {
        this.crashCount++;
      }
    });

    if (coreProcess.stderr) {
      coreProcess.stderr.on('data', (data: Buffer): void => {
        this.outputChannel.appendLine(`CoreBridge stderr: ${data.toString()}`);
      });
    }

    this.responseReader = readline.createInterface({
      input: coreProcess.stdout!,
      terminal: false
    });

    this.responseReader.on('line', (line: string): void => {
      if (this.pendingResolve) {
        const resolve: (line: string) => void = this.pendingResolve;
        this.pendingResolve = null;
        resolve(line);
      }
    });

    this.process = coreProcess;
    this.crashCount = 0;
    this.consecutiveTimeouts = 0;
  }

  /** Ensures the process is running, restarting if necessary (crash recovery). */
  private ensureProcess(): void {
    if (this.process) {
      return;
    }

    if (this.crashCount >= MAX_CRASHES) {
      vscode.window.showWarningMessage(
        'Babel TCC: The translation engine is unstable. Check the Output panel for details.'
      );
      throw new Error('CoreBridge: process crashed too many times, refusing to restart');
    }

    this.outputChannel.appendLine('CoreBridge: restarting process after crash');
    this.startProcess();
  }

  /**
   * Sends a request to the persistent process and waits for a response.
   * This is the internal method — callers use invokeCore() which handles queueing.
   */
  private sendRequest(method: string, params: Record<string, unknown>): Promise<CoreResponse> {
    return new Promise<CoreResponse>((resolve, reject): void => {
      try {
        this.ensureProcess();
      } catch (err: unknown) {
        reject(err);
        return;
      }

      if (!this.process || !this.process.stdin) {
        reject(new Error('CoreBridge: process not available'));
        return;
      }

      const requestLine: string = JSON.stringify({ method, params }) + '\n';
      this.outputChannel.appendLine(`CoreBridge: invoking ${method}`);

      const timeoutHandle: NodeJS.Timeout = setTimeout((): void => {
        this.pendingResolve = null;
        this.consecutiveTimeouts++;

        if (this.consecutiveTimeouts >= 2 && this.process) {
          this.outputChannel.appendLine('CoreBridge: 2 consecutive timeouts, killing process');
          this.process.kill();
        }

        reject(new Error(`CoreBridge: timeout after ${this.timeoutMs}ms for method ${method}`));
      }, this.timeoutMs);

      this.pendingResolve = (line: string): void => {
        clearTimeout(timeoutHandle);
        this.consecutiveTimeouts = 0;

        try {
          const response: CoreResponse = JSON.parse(line) as CoreResponse;
          if (!response.success) {
            this.outputChannel.appendLine(`CoreBridge error: ${response.error}`);
            reject(new Error(`CoreBridge: ${response.error}`));
            return;
          }
          this.outputChannel.appendLine(`CoreBridge: ${method} completed successfully`);
          resolve(response);
        } catch (parseErr: unknown) {
          reject(new Error(`CoreBridge: failed to parse response - ${line}`));
        }
      };

      this.process.stdin.write(requestLine);
    });
  }

  /**
   * Enqueues a request to the serial queue. Each request waits for the previous one to complete.
   */
  public invokeCore(method: string, params: Record<string, unknown>): Promise<CoreResponse> {
    return new Promise<CoreResponse>((resolve, reject): void => {
      this.requestQueue = this.requestQueue.then((): Promise<void> => {
        return this.sendRequest(method, params).then(resolve, reject).then((): void => {});
      }).catch((): void => {});
    });
  }

  public async translateToNaturalLanguage(
    sourceCode: string,
    fileExtension: string,
    targetLanguage: string
  ): Promise<string> {
    const params: Record<string, string> = { sourceCode, fileExtension, targetLanguage };
    const response: CoreResponse = await this.invokeCore('TranslateToNaturalLanguage', params);
    return response.result;
  }

  public async translateFromNaturalLanguage(
    translatedCode: string,
    fileExtension: string,
    sourceLanguage: string
  ): Promise<string> {
    const params: Record<string, string> = { translatedCode, fileExtension, sourceLanguage };
    const response: CoreResponse = await this.invokeCore('TranslateFromNaturalLanguage', params);
    return response.result;
  }

  public async applyTranslatedEdits(
    originalCode: string,
    previousTranslatedCode: string,
    editedTranslatedCode: string,
    fileExtension: string,
    sourceLanguage: string
  ): Promise<string> {
    const params: Record<string, string> = {
      originalCode, previousTranslatedCode, editedTranslatedCode, fileExtension, sourceLanguage
    };
    const response: CoreResponse = await this.invokeCore('ApplyTranslatedEdits', params);
    return response.result;
  }

  public async getKeywordMap(
    fileExtension: string,
    targetLanguage: string
  ): Promise<Record<string, string>> {
    const params: Record<string, string> = { fileExtension, targetLanguage };
    const response: CoreResponse = await this.invokeCore('GetKeywordMap', params);
    const parsed: Record<string, string> = JSON.parse(response.result) as Record<string, string>;
    return parsed;
  }

  public async getIdentifierMap(
    targetLanguage: string
  ): Promise<Record<string, string>> {
    const params: Record<string, string> = { targetLanguage };
    const response: CoreResponse = await this.invokeCore('GetIdentifierMap', params);
    const parsed: Record<string, string> = JSON.parse(response.result) as Record<string, string>;
    return parsed;
  }

  public async validateSyntax(
    sourceCode: string,
    fileExtension: string
  ): Promise<ValidationResult> {
    const params: Record<string, string> = { sourceCode, fileExtension };
    const response: CoreResponse = await this.invokeCore('ValidateSyntax', params);
    const parsed: ValidationResult = JSON.parse(response.result) as ValidationResult;
    return parsed;
  }

  public async getSupportedLanguages(): Promise<string[]> {
    const response: CoreResponse = await this.invokeCore('GetSupportedLanguages', {});
    const parsed: string[] = JSON.parse(response.result) as string[];
    return parsed;
  }

  public resolveTranslationsPath(context: vscode.ExtensionContext): string {
    const config: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration('babel-tcc');
    const configuredPath: string = config.get<string>('translationsPath', '');

    if (configuredPath && fs.existsSync(configuredPath)) {
      this.outputChannel.appendLine(`CoreBridge: translations via setting: ${configuredPath}`);
      return configuredPath;
    }

    if (this.projectPath) {
      const workspaceParent: string = path.dirname(this.projectPath);
      const siblingPath: string = path.join(workspaceParent, 'babel-tcc-translations');

      if (fs.existsSync(siblingPath)) {
        this.outputChannel.appendLine(`CoreBridge: translations auto-detected: ${siblingPath}`);
        return siblingPath;
      }
    }

    const embeddedPath: string = path.join(context.extensionPath, 'translations');
    this.outputChannel.appendLine(`CoreBridge: using embedded translations: ${embeddedPath}`);
    return embeddedPath;
  }

  public dispose(): void {
    this.disposed = true;
    this.outputChannel.appendLine('CoreBridge: disposing');

    if (!this.process) {
      return;
    }

    if (this.process.stdin) {
      this.process.stdin.write(JSON.stringify({ method: 'quit' }) + '\n');
    }

    const killTimeout: NodeJS.Timeout = setTimeout((): void => {
      if (this.process) {
        this.outputChannel.appendLine('CoreBridge: force killing process (quit timeout)');
        this.process.kill();
      }
    }, DISPOSE_TIMEOUT_MS);

    this.process.on('exit', (): void => {
      clearTimeout(killTimeout);
    });
  }
}

File: packages/ide-adapters/intellij/src/main/kotlin/com/nfasylum/babel/intellij/services/CoreBridge.kt
package com.nfasylum.babel.intellij.services

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.extensions.PluginId
import com.nfasylum.babel.intellij.BabelPlugin
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.io.Writer
import java.nio.charset.StandardCharsets
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.thread
import kotlin.concurrent.withLock

/** Response envelope returned by the C# Core.Host over the JSON Lines protocol (DT-002). */
data class CoreResponse(
    val success: Boolean = false,
    val result: String = "",
    val error: String = "",
)

/** How to launch the Core.Host process: a command plus its arguments. */
data class LaunchSpec(val command: String, val args: List<String>)

/** Raised when the Core engine cannot satisfy a request (crash, timeout, or a Core-side error). */
class CoreBridgeException(message: String) : RuntimeException(message)

/**
 * Bidirectional byte transport to a running Core.Host process. Extracted as an
 * interface so unit tests can drive [CoreBridge] with a scripted in-memory fake
 * instead of spawning the real .NET binary.
 */
interface CoreTransport {
    fun writeLine(line: String)

    /** Blocks until a full response line is available; returns null at end-of-stream. */
    fun readLine(): String?

    fun isAlive(): Boolean

    fun close()
}

/**
 * Kotlin port of the VS Code extension's `CoreBridge` (see
 * `packages/ide-adapters/vscode/src/services/coreBridge.ts`).
 *
 * Keeps a single long-lived Core.Host subprocess and talks to it with newline
 * delimited JSON (DT-002). Requests are serialised under a lock so each response
 * line maps unambiguously to the request that produced it — the Core answers
 * exactly one line per request, in order. A background thread drains stdout into
 * a queue so reads can honour a timeout without blocking forever.
 *
 * Registered as an application-level service; the no-arg constructor is required
 * by the platform. Tests construct it directly and replace [transportFactory].
 */
@Service(Service.Level.APP)
class CoreBridge {
    private val log = Logger.getInstance(CoreBridge::class.java)
    private val gson = Gson()

    /** Per-request timeout. Mutable so tests can shrink it. */
    var timeoutMs: Long = DEFAULT_TIMEOUT_MS

    /** Explicit Core.Host path override (a native binary or a `.dll`); null uses the bundled binary. */
    var coreHostPath: String? = null

    /** Passed to Core.Host as `--translations`; null lets the Core use its default. */
    var translationsPath: String? = null

    /** Passed to Core.Host as `--project`; null omits it. Enables identifier maps. */
    var projectPath: String? = null

    /** Seam for tests: builds a transport for a resolved [LaunchSpec]. Defaults to a real subprocess. */
    var transportFactory: (LaunchSpec) -> CoreTransport = ::spawnProcess

    private val lock = ReentrantLock()
    private val responses = LinkedBlockingQueue<String>()

    @Volatile
    private var transport: CoreTransport? = null

    /** Bumped on every (re)start; the reader thread uses it to ignore output from a superseded process. */
    @Volatile
    private var generation = 0

    private var restartCount = 0
    private var consecutiveTimeouts = 0

    @Volatile
    private var disposed = false

    // --- lifecycle --------------------------------------------------------

    /** Starts the Core.Host process if it is not already running. Idempotent. */
    fun start() {
        lock.withLock { ensureStarted() }
    }

    private fun ensureStarted() {
        val current = transport
        if (current != null && current.isAlive()) {
            return
        }
        if (restartCount >= MAX_CRASHES) {
            throw CoreBridgeException(
                "Core engine is unstable: refusing to restart after $MAX_CRASHES failed attempts.",
            )
        }
        startProcess()
    }

    private fun startProcess() {
        val spec = resolveLaunch()
        log.info("CoreBridge: starting Core.Host (${spec.command})")
        val newTransport = transportFactory(spec)
        responses.clear()
        val myGeneration = ++generation
        transport = newTransport
        restartCount++
        thread(name = "babel-core-reader-$myGeneration", isDaemon = true) {
            readLoop(newTransport, myGeneration)
        }
        log.info("CoreBridge started")
    }

    private fun readLoop(source: CoreTransport, myGeneration: Int) {
        try {
            while (generation == myGeneration) {
                val line = source.readLine() ?: break
                if (generation == myGeneration) {
                    responses.offer(line)
                }
            }
        } catch (e: Exception) {
            if (generation == myGeneration && !disposed) {
                log.warn("CoreBridge: reader thread stopped: ${e.message}")
            }
        }
    }

    /** Sends the quit sentinel and tears down the process. Safe to call more than once. */
    fun stop() {
        lock.withLock {
            disposed = true
            generation++
            transport?.let {
                try {
                    it.close()
                } catch (e: Exception) {
                    log.warn("CoreBridge: error during shutdown: ${e.message}")
                }
            }
            transport = null
        }
    }

    // --- request dispatch -------------------------------------------------

    /**
     * Sends one request and returns its [CoreResponse]. Serialised: concurrent
     * callers are queued so responses never interleave.
     */
    fun invoke(method: String, params: Map<String, String>): CoreResponse {
        lock.withLock {
            ensureStarted()
            val active = transport ?: throw CoreBridgeException("Core process unavailable")

            val request = JsonObject().apply {
                addProperty("method", method)
                add("params", gson.toJsonTree(params))
            }
            // Drop any late line left behind by a previously timed-out request so
            // ordering stays sound for this one.
            responses.clear()
            active.writeLine(gson.toJson(request))

            val line = responses.poll(timeoutMs, TimeUnit.MILLISECONDS)
            if (line == null) {
                onTimeout(method)
            }

            val response = parseResponse(line)
            consecutiveTimeouts = 0
            restartCount = 0
            if (!response.success) {
                throw CoreBridgeException("Core error for '$method': ${response.error}")
            }
            return response
        }
    }

    private fun onTimeout(method: String): Nothing {
        consecutiveTimeouts++
        // Kill the process so its stream is clean; the next call restarts fresh.
        generation++
        try {
            transport?.close()
        } catch (_: Exception) {
        }
        transport = null
        throw CoreBridgeException("Timeout after ${timeoutMs}ms for method '$method'")
    }

    private fun parseResponse(line: String): CoreResponse {
        return try {
            gson.fromJson(line, CoreResponse::class.java)
                ?: throw CoreBridgeException("Empty response from Core")
        } catch (e: Exception) {
            throw CoreBridgeException("Failed to parse Core response: $line")
        }
    }

    // --- typed API (mirrors the C# Program.cs method names) ---------------

    fun translateToNaturalLanguage(sourceCode: String, fileExtension: String, targetLanguage: String): String =
        invoke(
            "TranslateToNaturalLanguage",
            mapOf("sourceCode" to sourceCode, "fileExtension" to fileExtension, "targetLanguage" to targetLanguage),
        ).result

    fun translateFromNaturalLanguage(translatedCode: String, fileExtension: String, sourceLanguage: String): String =
        invoke(
            "TranslateFromNaturalLanguage",
            mapOf("translatedCode" to translatedCode, "fileExtension" to fileExtension, "sourceLanguage" to sourceLanguage),
        ).result

    fun applyTranslatedEdits(
        originalCode: String,
        previousTranslatedCode: String,
        editedTranslatedCode: String,
        fileExtension: String,
        sourceLanguage: String,
    ): String = invoke(
        "ApplyTranslatedEdits",
        mapOf(
            "originalCode" to originalCode,
            "previousTranslatedCode" to previousTranslatedCode,
            "editedTranslatedCode" to editedTranslatedCode,
            "fileExtension" to fileExtension,
            "sourceLanguage" to sourceLanguage,
        ),
    ).result

    /** Returns the translated-keyword -> original-keyword map for a language pair. */
    fun getKeywordMap(fileExtension: String, targetLanguage: String): Map<String, String> {
        val json = invoke(
            "GetKeywordMap",
            mapOf("fileExtension" to fileExtension, "targetLanguage" to targetLanguage),
        ).result
        return parseStringMap(json)
    }

    /** Returns the translated-identifier -> original-identifier map for the project. */
    fun getIdentifierMap(targetLanguage: String): Map<String, String> {
        val json = invoke("GetIdentifierMap", mapOf("targetLanguage" to targetLanguage)).result
        return parseStringMap(json)
    }

    fun getSupportedLanguages(): List<String> {
        val json = invoke("GetSupportedLanguages", emptyMap()).result
        return JsonParser.parseString(json).asJsonArray.map { it.asString }
    }

    private fun parseStringMap(json: String): Map<String, String> {
        val obj = JsonParser.parseString(json).asJsonObject
        return obj.entrySet().associate { (k, v) -> k to v.asString }
    }

    // --- launch resolution ------------------------------------------------

    /**
     * Resolves how to start Core.Host, mirroring the VS Code dual-track logic
     * (DT-010): prefer an explicit override, then the bundled native binary,
     * then a bundled `.dll` run via `dotnet`.
     */
    fun resolveLaunch(): LaunchSpec {
        val extraArgs = buildList {
            translationsPath?.let { add("--translations"); add(it) }
            projectPath?.let { add("--project"); add(it) }
        }

        coreHostPath?.let { path ->
            val file = File(path)
            if (file.exists()) {
                return if (file.extension.equals("dll", ignoreCase = true)) {
                    LaunchSpec("dotnet", listOf(file.absolutePath) + extraArgs)
                } else {
                    ensureExecutable(file)
                    LaunchSpec(file.absolutePath, extraArgs)
                }
            }
        }

        bundledHostFile(nativeBinaryName())?.let { native ->
            ensureExecutable(native)
            return LaunchSpec(native.absolutePath, extraArgs)
        }
        bundledHostFile(HOST_DLL_NAME)?.let { dll ->
            return LaunchSpec("dotnet", listOf(dll.absolutePath) + extraArgs)
        }

        // Last resort: assume the .NET runtime and the dll are discoverable.
        return LaunchSpec("dotnet", listOf(HOST_DLL_NAME) + extraArgs)
    }

    /** Locates a file under the installed plugin's `bin/` directory, or null if absent. */
    private fun bundledHostFile(name: String): File? {
        return try {
            val plugin = PluginManagerCore.getPlugin(PluginId.getId(BabelPlugin.PLUGIN_ID)) ?: return null
            val candidate = plugin.pluginPath.resolve("bin").resolve(name).toFile()
            if (candidate.exists()) candidate else null
        } catch (e: Exception) {
            null
        }
    }

    private fun ensureExecutable(file: File) {
        if (isWindows()) return
        try {
            file.setExecutable(true, false)
        } catch (e: Exception) {
            log.warn("CoreBridge: could not set executable bit on ${file.path}: ${e.message}")
        }
    }

    /** Default transport: spawn the real process and wrap its streams. */
    private fun spawnProcess(spec: LaunchSpec): CoreTransport {
        val process = ProcessBuilder(listOf(spec.command) + spec.args)
            .redirectErrorStream(false)
            .start()
        return ProcessTransport(process, gson)
    }

    companion object {
        const val DEFAULT_TIMEOUT_MS: Long = 10_000
        const val MAX_CRASHES: Int = 3
        const val HOST_DLL_NAME: String = "MultiLingualCode.Core.Host.dll"

        private fun isWindows(): Boolean =
            System.getProperty("os.name").orEmpty().startsWith("Windows", ignoreCase = true)

        private fun nativeBinaryName(): String =
            if (isWindows()) "MultiLingualCode.Core.Host.exe" else "MultiLingualCode.Core.Host"
    }
}

/** Real transport backed by a [Process], reading/writing UTF-8 JSON Lines. */
private class ProcessTransport(
    private val process: Process,
    private val gson: Gson,
) : CoreTransport {
    private val writer: Writer = OutputStreamWriter(process.outputStream, StandardCharsets.UTF_8)
    private val reader: BufferedReader = BufferedReader(InputStreamReader(process.inputStream, StandardCharsets.UTF_8))

    override fun writeLine(line: String) {
        writer.write(line)
        writer.write("\n")
        writer.flush()
    }

    override fun readLine(): String? = reader.readLine()

    override fun isAlive(): Boolean = process.isAlive

    override fun close() {
        try {
            writer.write(gson.toJson(mapOf("method" to "quit")))
            writer.write("\n")
            writer.flush()
        } catch (_: Exception) {
            // process may already be gone
        }
        if (!process.waitFor(DISPOSE_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
            process.destroyForcibly()
        }
    }

    private companion object {
        const val DISPOSE_TIMEOUT_MS: Long = 2_000
    }
}
